Signal/src/views/OWSSystemMessageCell.m

```diff
@@ -0,0 +1,205 @@
+#import "UIColor+OWS.h"
+#import "UIFont+OWS.h"
+#import "UIView+OWS.h"
+#import <JSQMessagesViewController/UIView+JSQMessages.h>
+#import <SignalServiceKit/TSCall.h>
+#import <SignalServiceKit/TSErrorMessage.h>
+#import <SignalServiceKit/TSInfoMessage.h>
+
+@interface OWSSystemMessageCell ()
+
+@property (nonatomic) UIImageView *imageView;
+@property (nonatomic) UILabel *titleLabel;
+
+@end
+
+#pragma mark -
+
+@implementation OWSSystemMessageCell
+
++ (NSString *)cellReuseIdentifier
+{
+    return NSStringFromClass([self class]);
+}
+
+- (void)configure
+{
+    self.backgroundColor = [UIColor whiteColor];
+
+    if (!self.titleLabel) {
+        self.imageView = [UIImageView new];
+        [self.contentView addSubview:self.imageView];
+
+        self.titleLabel = [UILabel new];
+        self.titleLabel.textColor = [UIColor colorWithRGBHex:0x403e3b];
+        self.titleLabel.font = [OWSSystemMessageCell titleFont];
+        self.titleLabel.numberOfLines = 0;
+        self.titleLabel.lineBreakMode = NSLineBreakByWordWrapping;
+        [self.contentView addSubview:self.titleLabel];
+    }
+
+    UIImage *icon = [self iconForInteraction:self.interaction];
+    self.imageView.image = [icon imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
+    self.imageView.tintColor = [self iconColorForInteraction:self.interaction];
+    self.titleLabel.text = [OWSSystemMessageCell titleForInteraction:self.interaction];
+    self.titleLabel.textColor = [self textColorForInteraction:self.interaction];
+
+    [self setNeedsLayout];
+}
+
+- (UIColor *)textColorForInteraction:(TSInteraction *)interaction
+{
+    return [UIColor colorWithRGBHex:0x303030];
+}
+
+- (UIColor *)iconColorForInteraction:(TSInteraction *)interaction
+{
+    // "Phone", "Shield" and "Hourglass" icons have a lot of "ink" so they
+    // are less dark for balance.
+    return [UIColor colorWithRGBHex:0x404040];
+}
+
+- (UIImage *)iconForInteraction:(TSInteraction *)interaction
+{
+    UIImage *result = nil;
+
+    if ([interaction isKindOfClass:[TSErrorMessage class]]) {
+        switch (((TSErrorMessage *)self.interaction).errorType) {
+            case TSErrorMessageInvalidKeyException:
+            case TSErrorMessageNonBlockingIdentityChange:
+            case TSErrorMessageWrongTrustedIdentityKey:
+            case TSErrorMessageMissingKeyId:
+                result = [UIImage imageNamed:@"system_message_security"];
+                break;
+            case TSErrorMessageNoSession:
+            case TSErrorMessageInvalidMessage:
+            case TSErrorMessageDuplicateMessage:
+            case TSErrorMessageInvalidVersion:
+            case TSErrorMessageUnknownContactBlockOffer:
+                result = [UIImage imageNamed:@"system_message_info"];
+                break;
+        }
+    } else if ([interaction isKindOfClass:[TSInfoMessage class]]) {
+        switch (((TSInfoMessage *)self.interaction).messageType) {
+            case TSInfoMessageUserNotRegistered:
```

Comment:
nit: any reason to have this separate from the `system_message_info` in the next block?